### DISCUSSION TITLE
Concurrency group does not support environmental variables?

### DISCUSSION THREAD

**Discussion by mikeschinkel** [community/community#26308] [Actions]
<p>Just checking to see if others can confirm or refute my analysis here.</p>
<p><strong>It seems that concurrency group does not support environmental variables</strong>?</p>
<p>Given this segment in one of my workflow files:</p>
<pre><code class="lang-auto">env:
   EXTERNAL_SERVER_IP: &lt;redacted&gt;
concurrency:
   group: my-build-${{ env.EXTERNAL_SERVER_IP }}-${{ github.ref_name }}
</code></pre>
<p>I get this Annotation error:</p>
<blockquote>
<p><a> <strong>Invalid workflow file:</strong> .github/workflows/ci_push.yaml#L18</a><br>
<code>The workflow is not valid. .github/workflows/ci_push.yaml (Line: 18, Col: 14): Unrecognized named-value: 'env'. Located at position 1 within expression: env.EXTERNAL_SERVER_IP</code></p>
</blockquote>
<p>I would like to use environment variables to stay D.R.Y. as I also use that variable elsewhere, but it seems maybe the GitHub Actions developers have to explicit implement every feature for every other feature, i.e. environment vars within concurrency, in this case?</p>
<p>Can anyone confirm this, or help me understand what I might be doing wrong?</p>
<p>And if in fact concurrency does not currently support environmental variables, what is the appropriate repo where I can submit an issue and/or maybe even a pull request?</p>
<p>Thanks in advance.</p>

**Comment by airtower-luna** [comment]
<p>This is documented in the <a href="https://docs.github.com/en/actions/learn-github-actions/contexts#env-context">description of the <code>env</code> context</a>:</p>
<blockquote>
<p>You can use the <code>env</code> context in the value of any key in a step except for the <code>id</code> and <code>uses</code> keys.</p>
</blockquote>
<p><code>concurrency</code> is not part of a step, so <code>env</code> isn’t available.</p>

**Comment by mikeschinkel** [comment]
<p>Thank you for the reply and explanation.</p>
<p>Is there any other way to define constants that would be in scope throughout the entirety of the YAML file?  <code>github.event.inputs.*</code> seems to have full YAML scope  for <code>workflow_dispatch</code>, but inputs do not seem to work for <code>on:push</code>.</p>
<p>-Mike</p>

**Comment by airtower-luna** [comment]
<p>For <a href="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency">workflow-level <code>concurrency</code></a> the documentation says that only the <code>github</code> context is available, so no. For <a href="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idconcurrency">job-level <code>concurrency</code></a> I suppose you could use another job to create an output and use that. :sweat_smile:</p>

**Comment by mikeschinkel** [comment]
<p>Thanks again, but unfortunately that does not sounds like a great option. But then your emoticon signaled that.</p>
<p>Do you mind if I ask my seconds question again from my original post?</p>
<blockquote>
<p><em>What is the appropriate repo where I can submit an issue and/or maybe even a pull request?</em></p>
</blockquote>

**Comment by airtower-luna** [comment]
<p>Considering that this is about the workflow syntax itself, I guess the Actions section of the community organization discussions would be the right place: <a href="https://github.com/orgs/github-community/discussions/categories/actions-and-packages" class="inline-onebox">Discussions · github-community · GitHub</a></p>

**Comment by mikeschinkel** [comment]
<p>Thank you for the reply!</p>

**Comment by nicholasbergesen** [comment]
<p>Hi <a class="mention" href="/mikeschinkel">@mikeschinkel</a></p>
<p>You could make use of reusable workflows to achieve what you’re trying to do.</p>
<p>caller.yml</p>
<pre><code class="lang-auto">on:
  push:
jobs:
  call_job:
    uses: ./.github/workflows/called.yml
    with:
      EXTERNAL_SERVER_IP: &lt;redacted&gt;
</code></pre>
<p>called.yml</p>
<pre><code class="lang-auto">on:
  workflow_call:
    inputs:
      EXTERNAL_SERVER_IP:
        type: string
        required: true
jobs:
  do_job:
    runs-on: ubuntu-latest
    concurrency:
      group: my-build-${{ inputs.EXTERNAL_SERVER_IP }}-${{ github.ref_name }}
    steps:
      - run: echo step
</code></pre>

**Comment by mikeschinkel** [comment]
<p>Hi <a class="mention" href="/nicholasbergesen">@nicholasbergesen</a>  —</p>
<p>Thank you for taking the time to offer a suggestion.</p>
<p>The irony of your suggestion, though, is that is exactly what I <strong><em>was</em></strong> doing prior to running into the unworkable limitations of reusable workflows and thus had to unravel that work and just go back to independent workflow files. And it was there I found myself trying to use environment variables as an alternative to the <code>input</code> values I had been using.</p>
<p>Seems I am damned if I do, damned if I don’t!  :cry:</p>
<p>So I made this <a href="https://github.com/orgs/github-community/discussions/18115" rel="noopener nofollow ugc">feature request</a>.</p>

**Comment by nicholasbergesen** [comment]
<p>Hi <a class="mention" href="/mikeschinkel">@mikeschinkel</a></p>
<p>An alternative workaround could be to use secrets. Add a secret in the repo/org and then reference it in the workflow using ${{ secrets.EXTERNAL_SERVER_IP }}. This would make the value accessible throughout the workflow similar to the behavior of <code>const</code> you proposed. It’s not the most elegant solution but it could give you the outcome you’re looking for.</p>

**Comment by mikeschinkel** [comment]
<p>Hi <a class="mention" href="/nicholasbergesen">@nicholasbergesen</a> —</p>
<p>Thank you again for the suggestion. Yes, secrets are certainly a creative stand-in for constants.</p>
<p>Unfortunately for my use-case they are unworkable for a few reasons:</p>
<ol>
<li>
<p>I specifically need the IP address of a vCenter server which I used to compose URLs to that vCenter server in the logs for use by the workflow user. If I used a secret it would redact that information and replace with asterisks thus making the URLs unusable and making it much more work for workflow users who need to visit the appropriate vCenter server after the workflow runs.</p>
</li>
<li>
<p>Also my work is at a large technology company and GitHub administration is managed by a DevOps department, not by me.  I do not have admin permissions to add secrets and adding non-secret secrets just for developer “convenience” is not something that team wants to manage and keep track of given the huge humber of secrets they already maintain and (my guess) especially since GitHub Secrets don’t have an associated description field to help inform the purpose of secrets for future maintainers…</p>
</li>
</ol>
<p>So yes, I agree secrets are a great workaround for constants when someone 1.) does not need to see the value while or after running a workflow and 2.) they have admin permissions to add and manage secrets themselves.</p>
<p>Unfortunately, for my use-case however, using secrets ends up being a non-starter. :cry:</p>
<p>-Mike</p>

**Comment by mikeschinkel** [comment, marked as answer]
<p>Hi <a class="mention" href="/nicholasbergesen">@nicholasbergesen</a> —</p>
<p>Thanks to your persistence in trying to identify a workaround for me you inspired me to search the docs to see if there were any other potentials solutions to address this need that actually fit my use-case.  And lo-and-behold, there is!</p>
<p>I was able to create a <strong>matrix strategy</strong> with one option like the following workflow snippet, and it worked!</p>
<pre><code class="lang-auto">strategy:
  matrix:
    EXTERNAL_SERVER_IP: [&lt;redacted&gt;]
concurrency:
  group: my-build-${{ matrix.EXTERNAL_SERVER_IP }}-${{ github.ref_name }}
</code></pre>
<p>Just thought I would close the loop on this for the benefit of anyone in the future who has the same need and googles for a solution.</p>
<p>Still, adding a <strong>constant</strong> feature to Github Action workflows would be much better as a <code>const</code>'s existence in a workflow would correctly imply why it is being used whereas a <code>matrix</code> might confuse future maintainers who miss  any comments explaining why a one-element matrix is being used.</p>

**Reply by julkwiec** [reply]
@mikeschinkel Thank you for posting this! This is probably the only documented solution to this issue :)

**Comment by nicholasbergesen** [comment]
<p>Hi <a class="mention" href="/mikeschinkel">@mikeschinkel</a></p>
<p>That’s a really creative solution, it never crossed my mind to do that. I’m glad you’ve found a way forward!</p>
<p>Thank you for sharing it.</p>
